usb/ehci/ehci.c

```diff
@@ -0,0 +1,737 @@
+		return NULL;
+
+	qtd->dt = datax;
+	qtd->pid = token;
+
+	qtd->next.terminate = 1;
+	qtd->altNext.terminate = 1;
+	qtd->active = 1;
+	qtd->errorCounter = 3;
+
+	if (buffer != NULL) {
+		qtd->offset = (uintptr_t)va2pa(buffer) & (EHCI_PAGE_SIZE - 1);
+		qtd->page0 = va2pa(buffer) >> 12;
+		offs = min(EHCI_PAGE_SIZE - qtd->offset, *size);
+		bytes += offs;
+		buffer += offs;
+
+		for (i = 1; i < 5 && bytes != *size; i++) {
+			qtd->buffers[i].page = va2pa(buffer) >> 12;
+			offs = min(*size - bytes, EHCI_PAGE_SIZE);
+			/* If the data does not fit one qtd, don't leave a trailing short packet */
+			if (i == 4 && bytes + offs < *size)
+				offs = (((bytes + offs) / maxpacksz) * maxpacksz) - bytes;
+
+			bytes += offs;
+			buffer += offs;
+		}
+
+		qtd->bytesToTransfer = bytes;
+		*size -= bytes;
+	}
+
+	return qtd;
+}
+
+
+static struct qh *ehci_allocQh(struct qh_node *node, usb_pipe_t *pipe)
+{
+	struct qh *qh;
+
+	if ((qh = usb_alloc(sizeof(struct qh))) == NULL)
+		return NULL;
+
+	qh->horizontal.terminate = 1;
+	qh->devAddr = pipe->dev->address;
+	qh->epSpeed = pipe->dev->speed;
+	qh->dt = pipe->type == usb_transfer_control ? 1 : 0;
+
+	qh->currentQtd.terminate = 1;
+	qh->transferOverlay.next.terminate = 1;
+
+	qh->maxPacketLen = pipe->maxPacketLen;
+	qh->ep = pipe->num;
+	qh->nakCountReload = 3;
+
+	if (pipe->type == usb_transfer_interrupt) {
+		if (node->qh->epSpeed == usb_high_speed) {
+			node->period = ((1 << (pipe->interval - 1))) >> 3;
+			/* Assume, that for 1-8 microframes period, we send it every microframe */
+			if (node->period == 0)
+				node->period = 1;
+		}
+		else {
+			node->period = 1;
+			while (node->period * 2 < pipe->interval)
+				node->period *= 2;
+		}
+		fprintf(stderr, "ehci: allocQh period: %d\n", node->period);
+	}
+
+	if (pipe->type == usb_transfer_control && pipe->dev->speed != usb_high_speed)
+		qh->ctrlEp = 1;
+
+	node->last = &qh->transferOverlay;
+	node->qh = qh;
+
+	return qh;
+}
+
+
+static void ehci_allocPeriodicBandwidth(ehci_t *ehci, struct qh_node *node)
+{
+	struct qh_node *tmp;
+	unsigned int i, n, best;
+	unsigned int ucnt[8] = { 0 };
+
+	best = (unsigned)-1;
+	node->phase = 0;
+	node->uframe = 0xff;
+
+	/* Find the best periodicList index (phase) to begin Qh linking */
+	for (i = 0; i < node->period && i < EHCI_PERIODIC_SIZE; i++) {
+		n = 0;
+		/* Count Qhs linked to this periodic index */
+		tmp = ehci->periodicNodes[i];
+		while (tmp != NULL) {
+			n++;
+			tmp = tmp->next;
+		}
+
+		if (n < best) {
+			best = n;
+			node->phase = i;
+		}
+	}
+	fprintf(stderr, "ehci: allocPeriodicBandwidth period: %d phase: %d\n", node->period, node->phase);
+
+	/* Find the best microframe in a frame. For periods equal to 1, send it every microframe */
+	if (node->qh->epSpeed == usb_high_speed && node->period > 1) {
+		tmp = ehci->periodicNodes[node->phase];
+		while (tmp != NULL) {
+			if (tmp->uframe != 0xff)
+				ucnt[tmp->uframe]++;
+			tmp = tmp->next;
+		}
```

Comment:
`for (tmp = ehci->periodicNodes[node->phase]; tmp != NULL; tmp = tmp->next) `perhaps?